packages/ui/src/providers/ListQuery/index.tsx

```diff
@@ -165,7 +172,13 @@ export const ListQueryProvider: React.FC<ListQueryProps> = ({
 
       hasSetInitialParams.current = true
     }
-  }, [defaultSort, defaultLimit, router, modifySearchParams, currentQuery])
+
+    // reset to first page if there are no results on the current page
+    if (data.docs.length === 0 && data.totalDocs > 0) {
+      currentQuery.page = '1'
+      router.replace(`?${qs.stringify(currentQuery)}`)
+    }
```

Comment:
@JessChowdhury on closer look, I don't think this should go here. I think it might belong in the `refineData` function.

And you can check if `search` or `where` are in the incoming query and reset page to 1 else keep it as it is in the currentQuery.

```ts
const refineListData = React.useCallback(
  async (query: RefineOverrides) => {
    if (!modifySearchParams) return
    
    let pageQuery = 'page' in query ? query.page : currentQuery?.page
    // reset to first page if where or search is updated
    if ('where' in query || 'search' in query) pageQuery = 1
    
    const updatedPreferences: Record<string, unknown> = {}
    let updatePreferences = false
    
    if ('limit' in query) {
      updatedPreferences.limit = query.limit
      updatePreferences = true
    }
    if ('sort' in query) {
      updatedPreferences.sort = query.sort
      updatePreferences = true
    }
    if (updatePreferences && preferenceKey) {
      await setPreference(preferenceKey, updatedPreferences)
    }
    const params = {
      limit: 'limit' in query ? query.limit : currentQuery?.limit,
      page: pageQuery,
      search: 'search' in query ? query.search : currentQuery?.search,
      sort: 'sort' in query ? query.sort : currentQuery?.sort,
      where: 'where' in query ? query.where : currentQuery?.where,
    }
    router.replace(`${qs.stringify(params, { addQueryPrefix: true })}`)
  },
  [preferenceKey, modifySearchParams, router, setPreference, currentQuery],
)
```

Reply:
Initially this is what I went for, I wasn't sure if it was because this is in a callback but the query it was returning was the previous one. I.e. if I added filter, the query comes up null, I then adjust the filter, the query will give me the previous conditions... like it's always 1 behind.

Reply:
I would have to see it to know what you mean. But `query` should just be what is passed in from the other functions i.e. `handleWhereChange`

Reply:
Disregard what I said here ^ this was happening because I was changing the page query right at the bottom of the callback previously. I've tested this and it works, will push this change now.